core/Archive/Purger.php

```diff
@@ -0,0 +1,228 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Archive;
+
+use Piwik\ArchiveProcessor\Rules;
+use Piwik\Config;
+use Piwik\DataAccess\ArchiveTableCreator;
+use Piwik\DataAccess\Model;
+use Piwik\Date;
+use Piwik\Db;
+use Piwik\Log;
+use Piwik\Piwik;
+use Piwik\Plugins\CoreAdminHome\Tasks\ArchivesToPurgeDistributedList;
+
+/**
+ * Service that purges temporary, error-ed, invalid and custom range archives from archive tables.
+ *
+ * Temporary archives are purged if they were archived before a specific time. The time is dependent
+ * on whether browser triggered archiving is enabled or not.
+ *
+ * Error-ed archives are purged w/o constraint.
+ *
+ * Invalid archives are purged if a new, valid, archive exists w/ the same site, date, period combination.
+ * Archives are marked as invalid via Piwik\Archive\Invalidator.
+ */
+class Purger
+{
+    /**
+     * @var Model
+     */
+    private $model;
+
+    /**
+     * Date threshold for purging custom range archives. Archives that are older than this date
+     * are purged unconditionally from the requested archive table.
+     *
+     * @var Date
+     */
+    private $purgeCustomRangesOlderThan;
+
+    /**
+     * Date to use for 'yesterday'. Exists so tests can override this value.
+     *
+     * @var Date
+     */
+    private $yesterday;
+
+    /**
+     * Date to use for 'today'. Exists so tests can override this value.
+     *
+     * @var $today
+     */
+    private $today;
+
+    /**
+     * Date to use for 'now'. Exists so tests can override this value.
+     *
+     * @var int
+     */
+    private $now;
+
+    public function __construct(Model $model = null, Date $purgeCustomRangesOlderThan = null)
+    {
+        $this->model = $model ?: new Model();
```

Comment:
ah ok I wasn't sure where you were using the class. I've taken the habit of doing `StaticContainer::get('The\Class')` (i.e. always go through the container) even though that's ugly, but at least that forces things to go in the right way: to remove the ugly you'd have to use even more DI